### PR TITLE
Tool implementations

### PR DESCRIPTION
**Pull request by johnswanson**
Skeleton for `generate-insights`, plus pulls in all the tools from ai-service (figure it'll be easier to just copy the schemas over and then work here, though we'll need to make sure they stay up to date).

### CODE REVIEW

File: enterprise/backend/src/metabase_enterprise/metabot_v3/tools/generate_insights.clj [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+   [metabase.public-settings :as public-settings]
+   [metabase.util.malli :as mu]))
+
+(mu/defmethod metabot-v3.tools.interface/*invoke-tool* :metabot.tool/generate-insights
+  [_tool-name {what-for :for :as _arguments} context]
+  (let [[k id] (some #(find what-for %) [:metric_id :table_id :report_id :query_id])
+        entity-type (case k
+                      :metric_id "metric"
+                      :table_id "table"
+                      :report_id "card"
```
Comment (metamben) [inline enterprise/backend/src/metabase_enterprise/metabot_v3/tools/generate_insights.clj:13]:
```suggestion
                      (:metric_id :report_id) "question"
                      :table_id "table"
```
`"metric"` would try to x-ray a legacy metric. New (v2) metrics are stored as cards, so `"question"` technically works for them. Metrics are strange beasts, and I'm not sure the x-ray we get like that is the best, but I don't know of any better. It's probably an open product question how to handle them.

File: enterprise/backend/src/metabase_enterprise/metabot_v3/tools/generate_insights.clj [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+                      :query_id (throw (ex-info "Unhandled" {})))]
```
Comment (metamben) [inline enterprise/backend/src/metabase_enterprise/metabot_v3/tools/generate_insights.clj:14]: This should be handled by using `"adhoc"` as entity type and the base64 encoded `dataset_query` as ID.

According to my current [vision](https://www.notion.so/Metabot-toolbox-v1-15169354c90180cf92edd3a32f8fc62c?d=15969354c90180a998d4001caa436381&pvs=4#15169354c9018014a350ff62edc9e914) with some rudiments implemented in #51315, is that you can find the query by this ID in what I call the full history.

File: enterprise/backend/src/metabase_enterprise/metabot_v3/dummy_tools.clj [hunk 3]
```diff
@@ -100,6 +107,31 @@
...
+(defn- get-metric-details
+  [_ {:keys [metric_id]} context]
+  (let [details (card-details metric_id)
+        date-field-id (:id (first (filter #(= (:type %) "date") (:fields details))))
+        details' (some-> details
+                         (assoc :queryable_dimensions (map convert-field (visible-columns (:dataset_query (t2/select-one :model/Card metric_id))))
+                                :default_time_dimension_field_id date-field-id)
+                         (select-keys [:id :description :name :default_time_dimension_field_id :queryable_dimensions]))]
+    {:output (or details' "metric not found")
+     :context context}))
```
Comment (metamben) [inline enterprise/backend/src/metabase_enterprise/metabot_v3/dummy_tools.clj:119]: I have an alternative implementation of this in #51315. ATM I think that the other `get-...-details` tools should be re-written to that style too for uniformity, if we the field IDs generated there are good enough for the LLM.

**Review by metamben — COMMENTED** [review]
I think `generate-insights` and `get-report-details` are good, I'm not sure about `get-metric-details`. We'll have to merge this with #51315.

**Comment by Somtom** [comment]
Do we want to implement all tools in this single branch now or rather go step by step with individual PRs?

**Comment by metamben** [comment]
> Do we want to implement all tools in this single branch now or rather go step by step with individual PRs?

No, we are going to merge to `metabot-v3-second-try`.

**Pull request merged**